### PR TITLE
[AI-Assisted] test(thermo): add numerical model specification gate

### LINKED ISSUE #3792: Feature request: a per-model "spec" harness so a path that returns 0 / NaN / a placeholder instead of a real number fails CI
**Summary.** The four defects I filed and you closed — #3768 (the unreachable DIPPR-101 branch,
20 components returning 1e38–1e96 bar), #3769 (`SystemUNIFAC` throws, `SystemUNIFACpsrk` returns
NaN), #3770 (Wilson/UNIQUAC fugacity coefficients of 0 or NaN), #3771 (Antoine defaults on 63%
of `COMP.csv`, `acetone` returning 241 bar) — are four presentations of one disease: a
calculation path that returns a *plausible-looking number* (`0.0`, `NaN`, a placeholder
constant, a value 90 orders of magnitude off) instead of either a real value or an explicit
"not available". Java's `double` cannot tell the two apart, and neither careful review nor the
current test suite reliably catches a missing assignment, a dead branch, or a placeholder row.
This is a feature request for the *guard* that would catch them, in Java's native form — not a
request to revisit any of the four fixes, which are fine.

## The one disease

Each issue is a different route to the same outcome — a number that looks like a result and is
not:

| Issue | The number | The actual defect |
|---|---|---|
| #3768 | 1e38–1e96 bar | a dead `else if` arm; the dispatch label disagrees with the coefficients |
| #3769 | exception / NaN | two fields hold "the groups"; only one is populated and they desync |
| #3770 | 0.0 / NaN | a coefficient computed and dropped, or stubbed, so the field stays zero |
| #3771 | 241 bar for `acetone`; a flat constant | placeholder data indistinguishable from a real correlation |

The common thread is **absence**, and the way Java represents it. A vapour pressure that does
not exist (a permanent gas above `Tc`, an ion), an activity coefficient that was never written,
a branch that no row reaches — each is "not there", and each is represented by an ordinary
`double` (or a zero-length array). Nothing in the signature says "this may be absent", so
nothing forces a caller or a test to handle the absence. The value ships, and it is found only
when someone drives the model and happens to notice the number is absurd.

## Where I am coming from, and what is actually portable

I maintain [azoth](https://github.com/PatrickMockridge/Azoth), a chemical-engineering library
in Rust that ports much of NeqSim's physics. It catches these four classes mechanically, and I
want to be precise about which part of that is portable to a Java codebase, because the honest
answer is "almost all of it":

1. **A worked example per calculation.** Every model ships one concrete `(inputs → output)`
   pair, and CI asserts the implementation reproduces it to a tight tolerance. An i-pentane
   `getAntoineVaporPressure(298.15) == 0.918 bar` entry fails the moment the DIPPR branch is
   missed, whether or not anyone read the code. This is the highest-leverage rule and it is
   completely language-neutral.

2. **Absence is a value, not a number.** "No correlation available" is a distinct type — Rust's
   `Option`/`Result`, i.e. a sum type with no `null` and no `0.0`/`NaN` sentinel. Java has
   `OptionalDouble` and checked exceptions; the same rule applies.

3. **Physical invariants as tests.** `g == h − T·s`, `h == u + p·v`, activity coefficients
   positive, fugacity coefficients finite and positive. Cheap to write, and one assertion covers
   a whole family of "silently wrong" paths at once.

4. **Exhaustive dispatch.** Correlations and mixing rules are `enum`s, and a `switch` (or
   Rust `match`) over an enum with a missing arm is a compile error rather than a dead branch.

5. *(Far end, not proposed here.)* Machine-checked units and dimensions, with a few Lean
   theorems gating the unit constructions. That part is a property of Rust's type system plus a
   proof assistant, and its payoff is at the dimension boundary, not the value — so I am
   **not** suggesting NeqSim adopt it. The value-correctness above comes entirely from 1–4.

The point is not "rewrite in Rust". It is that 1–4 are each a discipline, and each can be
adopted in Java today, incrementally.

## The proposed feature

A **per-model spec harness**: a curated smoke-test catalog, a worked-example data table, and a
convention that absence is explicit. Three parts.

### Part 1 — the smoke-test catalog (catches #3770 and #3768)

A data table of `(system, method, bound)` entries, walked by one JUnit test. One entry per model
family — each `System*` subclass — driving it with a standard smoke-test mixture and asserting,
for every number the model publishes:

- `Double.isFinite(value)` — no NaN, no ±Infinity;
- `value > 0` for activity coefficients and fugacity coefficients (this alone fails #3770:
  `gamma == 0.0` is a perfectly finite `double`, and it is the *positivity* that exposes it);
- `0 <= value <= somePhysicalCeiling` for vapour pressures, where the ceiling is the
  component's critical pressure (this alone fails #3768's 1e38 bar and #3771's 241-bar
  `acetone`).

The table is curated rather than reflective-over-the-classpath, because a model is only
meaningfully "driven" when someone has said what it should produce — reflection can't write the
ceiling. This is azoth's "drive the thing and read the number" rule made mechanical: #3770's
`PhaseGEWilson` *looks* implemented — class, method, javadoc, all present — and only reading the
number shows `gamma == 0.0`.

### Part 2 — the worked-example table (catches all four)

A small data file of `(components, T, P, expected value)` rows, each naming a property and an
exact expected value, compared at a tight relative tolerance. This is the single most valuable
part. It turns "the model ran without throwing" into "the model ran and produced the right
number", which is the distinction every one of the four issues crossed:

- #3769's test asserted a component *count* and never read a group — a worked example reads the
  group;
- #3770's phase returned `0.0` silently — a worked example of `gamma` for
  n-butane/water fails;
- #3768 and #3771's absurd pressures are a worked example that disagrees by 30–90 orders of
  magnitude.

Each row needs an independent source for its expected value — a published datum, a handbook
value, or a second implementation. Where the second implementation exists (azoth ports much of
NeqSim and reproduces it to 1e-12 on several models), it can seed the table for free, and the
comparison works in both directions.

### Part 3 — absence that is a value (catches #3768 and #3771)

Two conventions, adopted gradually:

- **`OptionalDouble` (or a checked exception) for "no correlation available".** A permanent gas
  above `Tc`, or an ion, has no saturation vapour pressure. Returning `OptionalDouble.empty()`
  makes "not there" a distinct value that a caller must handle, instead of a `42.24 Pa` constant
  that is indistinguishable from a real result. #3771's placeholder rows become *visible* the
  moment the policy is "placeholders return empty".
- **An `enum` for the Antoine correlation type instead of a string label + `else if` chain.**
  `antoineLiqVapPresType` is currently a string, so #3768's dead DIPPR-101 arm — guarded by
  `Math.abs(AntoineE) > 1e-12` but placed after the `log` arm that every `E != 0` row matches —
  cannot be flagged by anything. A `switch` over an `AntoineForm` enum, compiled with
  exhaustiveness, makes a reachable-but-unmapped form a compiler error and a mapped-but-unused
  form a linter error.

## Why the current suite missed all four

Not a criticism of the tests — they miss them for a specific reason that the spec harness is
designed around:

- #3769's test asserts `getNumberOfComponents()` and never drives a group read, so a zero-length
  array passes.
- #3770's phase is *hollow*, not visibly broken: the class, the method, and the javadoc all
  exist; one assignment is missing. A test that only checks "did it throw" passes.
- #3768 and #3771 are invisible until the value is *evaluated and judged*. A dead branch throws
  nothing; a placeholder returns a number. Only a bound (Part 1) or an expected value (Part 2)
  sees them.

## Incremental path and cost

- **Part 2 is cheapest and catches the most.** A worked-example table with even a handful of
  rows per model family is a self-contained PR and would have failed #3768, #3770 and #3771 at
  the value level.
- **Part 1 is a curated list plus a loop.** It rides the same data file, with looser assertions
  (finite, positive, bounded) that need no external source.
- **Part 3 is the most invasive and can wait.** It touches public signatures
  (`getAntoineVaporPressure` returning `OptionalDouble` would break callers), so it belongs
  behind the other two or behind a deprecation cycle.

None of this changes the physics or the data, and none of it requires leaving Java. It is the
testing-and-typing half of what azoth mechanizes, restated in Java's own idioms.

## Explicitly not proposed

- No rewrite in Rust, no adoption of Lean or any proof assistant.
- No change to the four fixes already merged — they are correct; this is about making the
  *next* one of the same shape fail in CI rather than ship.

### PR DESCRIPTION
**Pull request by EvenSol**
## Scope
Refs #3792, Milestone A. This is the campaign's sole implementation PR. **Draft; focused native Java 8/21 CI passes. Broader CI and review remain pending.** No production equations/data/public APIs are changed.

Frozen concept: a curated test-side numerical specification gate, its initial worked examples, negative controls, CI and contributor guide. Stop boundary: complete-model inventory/physical family qualification (B), typed saturation API/Antoine enum (C), and production derivative repair #3798 are separate dependencies. This does not close #3792.

## Implementation
- Strict versioned 21-column TSV catalog with typed fixture/property/outcome registries; metadata, units, molar composition, temperature domain, finite references/nonnegative tolerances, uniqueness and required-case validation.
- Declared VALUE / UNAVAILABLE / UNSUPPORTED contracts. Unexpected NaN/exception fails; absence is never inferred from the observed answer.
- 31 sourced/analytical catalog cases: NIST acetone and i-pentane at three temperatures each; six nonideal Wilson component anchors and negative ln(gamma); classic UNIFAC/PSRK/UMR-PRU group contents/R and pure gamma; legitimate SRK/PR dilute-Z and zero ideal-H controls; missing/ion/supercritical saturation data; explicit UNIQUAC rejection.
- Read returned and stored gamma/ln-gamma/phi, exercise changed state/component order/reinitialization, and cover six analytical pressure-dispatch forms. Existing focused regression tests remain unchanged.
- Per-case TSV evidence, required-ID and adapter coverage checks, malformed catalog and numerical negative controls. This initial coverage is six concrete System drivers plus standalone component/phase paths, NOT all NeqSim models.
- Ordinary fast-suite discovery plus strict focused Java 8/21 CI. Fix data-only/test-resource-only path-filter gap. Artifact reports retained; no failIfNoTests=false or continue-on-error in the new gate.
- Developer guide/index covering schema, provenance, absence semantics, implementation-vs-physical tolerances and remaining coverage debt. Branch-protection requirements remain administrator policy.

## Validation on 2026-09-18
Base: `8a98ac18924ae6e24c6423ac426a257c4a54648e`.
Head: `73b4a1c95579971323ec983c7237c35756b2c529`.
Local staged tree exactly matches published tree `0c0064b992e2932b138a077c2b15ed022f86e3de`.

Environment: OpenJDK 17.0.20, Python 3.12.14, JUnit 6.1.3, Surefire 3.6.0; Maven bootstrapped using the prescribed helper. Native Temurin 8.0.504+1 and 21.0.12+1 execution also passed the added CI matrix: 70 tests each, zero failures/errors/skips. Maven build/test elapsed times were 48.359 s (8) and 51.176 s (21). These are actual CI results, separate from local Java 8 API-surface compilation.

Passed:
```text
./mvnw -q -DskipITs -Djacoco.skip=true "-Dtest=ModelSpec*Test,ComponentCorrelationSpecTest,ComponentAntoineVaporPressureTest,AntoineDataAvailabilityTest,GeActivityCoefficientPublicationTest,UnifacGroupSynchronizationTest" test
# 132 tests: 70 new + 62 existing; zero failures/errors/skips.
java com.sun.tools.javac.Main --release 8 -cp <production+test+dependency-classpath> -d <temporary-directory> <six new Java files>
# New test sources compile against the Java 8 API surface; not a native JDK 8 runtime claim.
./mvnw -q -Djacoco.skip=true "-Dtest=ModelSpec*Test,ComponentCorrelationSpecTest" -DfailIfNoTests=true -Dsurefire.failIfNoSpecifiedTests=true surefire:test
# Warmed focused execution: 70 tests, 5.34 s wall; no compilation included.
<python> devtools/run_spotless.py apply
<python> devtools/run_spotless.py check
<python> devtools/check_documentation_search.py
pre-commit run --all-files --hook-stage pre-commit
pre-commit run --all-files --hook-stage pre-push
git diff --cached --check
```
Spotless stable; direct and hook gates passed. Documentation audit: 713 Markdown pages + 1 standalone HTML. Unmodified-base focused regression baseline also passed (62 tests). No full-suite pass is claimed.

### Production mutation evidence (not just helper self-tests)
Each mutant was compiled into production classes, then `./mvnw -q -Djacoco.skip=true -Dtest=ModelSpecTest surefire:test` ran all 31 catalog cases. Each exited 1. Source was restored between mutants and the final 132-test run passed.

| Defect family | Temporary production mutation | Detected |
|---|---|---|
| #3768 dispatch | Disable DIPPR pressure dispatch branch | 3 failures |
| #3769 group synchronization | Leave component UNIFAC array empty | 9 errors |
| #3770 publication | Return computed Wilson gamma without assigning stored gamma | 7 failures |
| #3771 plausible placeholder | Return constant 0.306 bar for acetone | 3 failures |

NIST anchors are derived from cited correlations, never refreshed from NeqSim output. Acetone shares the published fitted formula (implementation check); i-pentane compares a distinct NIST fit with existing DIPPR data (1% comparison tolerance). Pure gamma=1 is explicitly documented as a limited control, not independent mixture validation.

## Known dependency and follow-up
A separate real production defect was reproduced: pow10KPa pressure derivative returns zero for a nonconstant available correlation (finite-difference slope 0.0153505673 bar/K). Filed #3798 with reproducer; not hidden by a passing pressure contract, and not repaired opportunistically here.

Next: inspect exact-head CI/review, finish any justified Milestone A follow-up in this PR, then proceed only after merge to the complete inventory/coverage gate and sourced family batches. No merge, auto-merge or ready-for-review request is authorized.

### Exact-head CI follow-up
Verified at head `73b4a1c95579971323ec983c7237c35756b2c529`: [Java 8 spec gate](https://github.com/equinor/neqsim/actions/runs/35386131861/job/105733464854) and [Java 21 spec gate](https://github.com/equinor/neqsim/actions/runs/35386131861/job/105733464916) pass. Changes filter, Javadoc, formatting, pre-commit and agent-skill checks pass. Broader build/security/fast/slow checks were still running at inspection. Mergeable without conflicts but blocked/draft; no reviews or unresolved review threads at inspection. Not READY TO MERGE.

### CODE REVIEW

**Comment by EvenSol** [comment]
The isolated production repair for #3798 is now in #3803 (head `b7f66792cd1362173a6238d2684bafbb98001203`). It implements the pow10KPa derivative in bar/K without changing the pressure scale or dispatch precedence. All 52 focused Antoine tests pass, including ten new cases; on unmodified production code, six derivative and three inverse-temperature cases fail.

Catalog integration remains here under #3792 once the repair is available. The prescribed fixture A = log10(2) + 6, B = 300, C = D = E = 0 at 300 K has pressure 2 bar and derivative `ln(10) / 150 = 0.01535056728662697 bar/K`. Please preserve this as a sourced analytical VALUE contract, with additional temperatures and explicit-label/nonzero-E precedence coverage. The repair's `ComponentPow10KPaVaporPressureTest` supplies finite-difference and round-trip regressions that can remain independent of the catalog.

#3798 is left open because its requested catalog entry is still pending this harness integration. The separate Wagner derivative path is unchanged.

File: src/test/java/neqsim/thermo/spec/ModelSpec.java [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+      require(amount > 0.0 && !components.containsKey(pair[0]), "invalid/duplicate component: " + entry);
+      components.put(pair[0], amount);
+      sum += amount;
+    }
+    require(Math.abs(sum - 1.0) <= 1e-12, "composition must sum to one mole");
+    temperature = finite(row[4]);
+    pressure = finite(row[5]);
+    require(temperature > 0.0 && pressure > 0.0, "T and absolute P must be positive");
+    phase = row[6];
+    componentIndex = Integer.parseInt(row[7]);
```
[the remark this answers is not in the document]
Comment (EvenSol) [reply src/test/java/neqsim/thermo/spec/ModelSpec.java:81]: Reviewed in the #3792 follow-up: NumberFormatException already extends IllegalArgumentException, which ModelSpec.read catches and wraps with the catalog line number while preserving the cause. No redundant catch is needed. PR #3869 adds five invalid-number cases covering componentIndex, temperature, reference and tolerance fields; all pass and explicitly assert `catalog line 3:` plus a NumberFormatException cause. The existing validation behavior is retained.

File: src/test/java/neqsim/thermo/spec/ModelSpec.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+      } catch (IllegalArgumentException ex) {
+        throw new IllegalArgumentException("catalog line " + number + ": " + ex.getMessage(), ex);
+      }
+    }
+    require(!cases.isEmpty(), "empty catalog");
+    return cases;
+  }
+
+  private static double finite(String value) {
+    double parsed = Double.parseDouble(value);
```
[the remark this answers is not in the document]
Comment (EvenSol) [reply src/test/java/neqsim/thermo/spec/ModelSpec.java:140]: Reviewed in the #3792 follow-up: NumberFormatException already extends IllegalArgumentException, which ModelSpec.read catches and wraps with the catalog line number while preserving the cause. No redundant catch is needed. PR #3869 adds five invalid-number cases covering componentIndex, temperature, reference and tolerance fields; all pass and explicitly assert `catalog line 3:` plus a NumberFormatException cause. The existing validation behavior is retained.
